### PR TITLE
feat: in world builder audio

### PR DESCRIPTION
**Pull request by Landeplage**
# What? 
Music and sound effects for the in-world builder.

# Why? 
It's a bit boring to use without audio.

### CODE REVIEW

**Comment by Landeplage** [comment]
Note: This branch is based off of the "feat/avatar-audio" branch.

File: unity-client/Assets/Scenes/InitialScene.unity [hunk 1]
```diff
@@ -38,7 +38,7 @@ RenderSettings:
   m_ReflectionIntensity: 1
   m_CustomReflection: {fileID: 8900000, guid: b4f78c510a8526547ab9c9946449480e, type: 3}
   m_Sun: {fileID: 943100645024490816}
-  m_IndirectSpecularColor: {r: 0.10252202, g: 0.24458984, b: 0.36489132, a: 1}
+  m_IndirectSpecularColor: {r: 0.00076876703, g: 0.8351885, b: 0.8361403, a: 1}
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scenes/InitialScene.unity]: Please, discard all the changes on the InitialScene they are not needed

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        DCL.Tutorial.TutorialController.i.OnTutorialDisabled -= OnTutorialDisabled;
+
+        for (int i = 0; i < builderInWorldModes.Length; i++) {
+            builderInWorldModes[i].OnEntityDeselected -= OnAssetDeselect;
+            builderInWorldModes[i].OnEntitySelected -= OnAssetSelect;
+        }
+    }
+
+    private void OnEnterEditMode() {
+        StartCoroutine(StartBuilderMusic());
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: We have a coroutine starter so we can handle all the coroutines from there, please use` CourutineStarter.Start` maybe you will need to add the dependency, don't hesitate to ask for help for this!

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    private void OnExitEditMode() {
+        eventBuilderExit.Play();
+        StartCoroutine(eventBuilderMusic.FadeOut(5f));
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: Can we move the 5f to a constant in the class?

Also, please use the `CourutineStarter`!

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    private void OnAssetSelect() {
+        AudioScriptableObjects.inputFieldUnfocus.Play(true);
+    }
+
+    private void OnAssetDeselect(DCLBuilderInWorldEntity entity, bool assetIsNew) {
+        if (assetIsNew || modeController.GetCurrentStateMode() == BIWModeController.EditModeState.FirstPerson) {
+            eventAssetPlace.Play();
+
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: What do you think about the use of the sounds even if the "asset" is not new? This way we always have feedback each time you deselect an item even in GodMode

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            if (entitiesOutOfBounds.Contains(entity.rootEntity.entityId)) {
+                eventBuilderOutOfBoundsPlaced.Play();
+            }
+        }
+    }
+
+    private void OnTutorialEnabled() {
+        StartCoroutine(eventBuilderMusic.FadeOut(3f));
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: this should happen only if we have the builder in world active right?, please use the boolean `isBuilderInWorldActivated` from `BuilderInWorldController` to check if it is active

Comment (D4rWiNSS) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: Also use the CoroutineStarter

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    private void OnTutorialDisabled() {
+        if (modeController.GetCurrentStateMode() != BIWModeController.EditModeState.Inactive)
+            StartCoroutine(StartBuilderMusic());
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: please use the boolean `isBuilderInWorldActivated` from `BuilderInWorldController` to check if it is active.
 
 Also use the CoroutineStarter

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    private IEnumerator StartBuilderMusic() {
+        yield return new WaitForSeconds(4f);
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: Can we move this to a constant please?

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+        if (modeController.GetCurrentStateMode() != BIWModeController.EditModeState.Inactive)
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: please use the boolean `isBuilderInWorldActivated` from `BuilderInWorldController` to check if it is active

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs [hunk 9]
```diff
@@ -22,8 +23,10 @@ public class BuilderInWorldMode : MonoBehaviour
 
     public ActionController actionController;
 
-    public event System.Action OnInputDone;
-    public event System.Action<BuildInWorldCompleteAction> OnActionGenerated;
+    public event Action OnInputDone;
+    public event Action<BuildInWorldCompleteAction> OnActionGenerated;
+    public event Action<DCLBuilderInWorldEntity, bool> OnEntityDeselected;
+    public event Action OnEntitySelected;
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs]: This action should be moved and called to the `BuilderInWorldEntityHandler`,  is his responsibility to handle this kind of selections

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs [hunk 10]
```diff
@@ -75,6 +86,8 @@ public virtual void SelectedEntity(DCLBuilderInWorldEntity selectedEntity)
         CenterGameObjectToEdit();
 
         BuilderInWorldUtils.CopyGameObjectStatus(editionGO, undoGO, false, false);
+
+        OnEntitySelected?.Invoke();
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs]: we should move this to entity handler

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs [hunk 11]
```diff
@@ -107,6 +120,8 @@ public virtual void EntityDeselected(DCLBuilderInWorldEntity entityDeselected)
             actionController.CreateActionEntityCreated(entityDeselected.rootEntity);
         }
 
+        OnEntityDeselected?.Invoke(entityDeselected, isNewObjectPlaced);
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs]: we should move this to entity handler

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/BuildModeHUD/Scripts/BuildModeInWorldLoadingAudioHandler.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        loadingView.OnHide -= OnLoadingViewHide;
+    }
+
+    private void OnEnable() {
+        AudioScriptableObjects.builderEnter.Play();
+    }
+
+    private void OnLoadingViewHide() {
+        AudioScriptableObjects.builderReady.Play();
+    }
```
Comment (D4rWiNSS) [inline unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/BuildModeHUD/Scripts/BuildModeInWorldLoadingAudioHandler.cs]: We can move this to the BuilderInWorldLoadingView directly, so we don't have another class to handle this

**Review by D4rWiNSS — CHANGES_REQUESTED** [review]
Good Job! These sounds add a lot to the builder in world, it sounds very good! 

I have added some feedback and made some changes to the PR, don't hesitate to contact me if you need help or if something is not clear enough!

**Comment by D4rWiNSS** [comment]
After testing the sounds this is my feedback, however don't implement this feedback in this PR, after we all agree on the feedback we can create another branch with another PR:

- Selection sound is too quiet compare to the music, sometimes the music will override the sound
- Error sound is identification enough? Maybe I am alone in this, but I think that the error sound doesn't sound like "error" enough 
- If the selected entity has been moved, we should use the "placed item" sound as well to give audio feedback 
- If we overpass the limits of the lands (triangles, textures...etc) the error sound doesn't sound
- While enter/exit from builder in world, differents sounds sound at the same time

**Comment by D4rWiNSS** [comment]
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!             DANGER           !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Please test this feature only in zone ("ropsten") , in the 140, 140 , now there is a save system that will save your progress

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs (hunk 2 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: Fixed

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs (hunk 3 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: Fixed

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs (hunk 5 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: Fixed

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs (hunk 6 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: Fixed

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs (hunk 7 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: Fixed

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs (hunk 8 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: Fixed

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs (hunk 9 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs]: Fixed

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs (hunk 10 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs]: Fixed

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs (hunk 11 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/States/BuilderInWorldMode.cs]: Fixed

File: unity-client/Assets/Scenes/InitialScene.unity (hunk 1 above)
Comment (Landeplage) [reply unity-client/Assets/Scenes/InitialScene.unity]: I'll try to keep this in mind going forward.

**Review by pravusjif — APPROVED** [review]

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/BuildModeHUD/Scripts/BuildModeInWorldLoadingAudioHandler.cs (hunk 12 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/BuildModeHUD/Scripts/BuildModeInWorldLoadingAudioHandler.cs]: Fixed

**Review by sandrade-dcl — APPROVED** [review]

File: unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs (hunk 4 above)
Comment (Landeplage) [reply unity-client/Assets/Scripts/MainScripts/DCL/Controllers/BuilderMode/BuilderInWorldAudioHandler.cs]: Implemented a "deselect" sound which plays when entity is deselected, except the first deselect after selecting a catalogue item (in which case the placement sound is played).

**Comment by Landeplage** [comment]
> 
> 
> After testing the sounds this is my feedback, however don't implement this feedback in this PR, after we all agree on the feedback we can create another branch with another PR:
> 
>     * Selection sound is too quiet compare to the music, sometimes the music will override the sound
> 
>     * Error sound is identification enough? Maybe I am alone in this, but I think that the error sound doesn't sound like "error" enough

I made the error sound a bit more error-sounding, and edited the selection sound slightly. It might still be too similar to the music. Maybe we can change it into a more "mechanical" sound later, in the same style as the other builder-specific sound effects.

**Review by D4rWiNSS — APPROVED** [review]

**Pull request merged**